blog/2023-02-20-introducing-the-dhis2-java-sdk.md

```diff
@@ -0,0 +1,174 @@
+
+* `transfer()` executes the HTTP call, that is, sends the request to the DHIS2 server.
+
+* `returnAs(...)` deserialises the HTTP response JSON body into a POJO of type `DataValueSet` since we are expecting a data value set from `https://source.dhis2.org/api/dataValueSets`.
+
+:::info
+By default, a `get(...)` call assumes that a single resource is being fetched. Fetching a list of resources from a collection endpoint is accomplished by appending `withPaging()` or `withoutPaging()` to `get(...)`. For instance, if we wanted to retrieve the list of organisation units in pages, then the method chain would be written as:
+
+```java
+Iterable<OrganisationUnit> orgUnits = sourceDhis2Client.get( "organisationUnits" )
+    .withPaging()
+    .transfer()
+    .returnAs( OrganisationUnit.class, "organisationUnits" );
+```
+
+`returnAs` requires an additional parameter when `withPaging()` or `withoutPaging()` is applied to the method chain. This parameter is the property name of the JSON array holding the resource items within the response. The JSON array name in this example is `organisationUnits` given that organisation units are going to be fetched. Notably, fetching a collection changes the method chain’s return signature to an iterable of the type specified in `returnAs(...)`, in this case, `Iterable<OrganisationUnit>`. When pagination is applied, the iterator is lazy: the iterator transparently fetches and feeds the subsequent page to its `next()` method after reaching the last item of the current page.
+:::
+
+The `sourceDhis2Client` method chain retrieving the data value set returns an object belonging to a `DataValueSet` _resource class_. Rather than attempting to parse raw JSON strings or reference unsafe generic _Maps_, a resource class such as `DataValueSet` provides a type-safe fluent view of the request/response's content to the application developer. Additionally, with the help of an IDE’s autocompletion, the programmer can explore which fields are available instead of digging into the Web API documentation. 
+
+Resource classes are located within the package `org.hisp.dhis.api.model.vX_X_X` where `X_X_X` is a variable informing us of the DHIS2 version that the classes are compatible with. Taking `DataValueSet` from our example, this class is located inside the package `org.hisp.dhis.api.model.v2_39_1`. The version no. `v2_39_1` means that this `DataValueSet` class is compatible with DHIS 2.39.1. If we decide to update the application to support a newer version of DHIS2, then we should bump the package version no. `v2_39_1` to match the new DHIS2 version. Any Web API breaking changes like the renaming of a JSON property would be reflected in the corresponding resource class. This in turn would cause the application to fail compilation should it be calling an accessor that is impacted by the breaking change. 
```

Comment:
```suggestion
Resource classes are located within the package `org.hisp.dhis.api.model.vX_X_X` where `X_X_X` is a variable informing us of the DHIS2 version that the classes are compatible with. 

Taking `DataValueSet` from our example, this class is located inside the package `org.hisp.dhis.api.model.v2_39_1`. 

The version no. `v2_39_1` means that this `DataValueSet` class is compatible with DHIS 2.39.1. If we decide to update the application to support a newer version of DHIS2, then we should bump the package version no. `v2_39_1` to match the new DHIS2 version. 

Any Web API breaking changes like the renaming of a JSON property would be reflected in the corresponding resource class. This in turn would cause the application to fail compilation should it be calling an accessor that is impacted by the breaking change. 
```